webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeConfigurationService.scala

```diff
@@ -0,0 +1,352 @@
+*
+* This file is part of Rudder.
+*
+* Rudder is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* In accordance with the terms of section 7 (7. Additional Terms.) of
+* the GNU General Public License version 3, the copyright holders add
+* the following Additional permissions:
+* Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+* Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU General
+* Public License version 3, when you create a Related Module, this
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.services.reports
+import com.normation.box._
+import com.normation.errors._
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.domain.logger.{ReportLogger, ReportLoggerPure}
+import com.normation.rudder.domain.policies.RuleId
+import com.normation.rudder.domain.reports.{NodeAndConfigId, NodeExpectedReports, NodeStatusReport}
+import com.normation.rudder.repository.{CachedRepository, FindExpectedReportRepository}
+import com.normation.rudder.services.nodes.NodeInfoService
+import com.normation.utils.Control.sequence
+import com.normation.zio._
+import net.liftweb.common._
+import org.joda.time._
+import zio._
+import zio.syntax._
+
+/**
+ * That service retrieve node configurations (nodeexpectedreports) from the expectedreportsjdbcrepository, unless its already in cache
+ * cache is driven by reporting serviceimpl
+ * init add all nodes, withtout anything attached
+ * initial setting of nodeexpectedreport is less prioritary than update
+ * deletion removes the entry
+ * if an entry exists but without anything, then it will query the database
+ * if an entry exists but wthout the right nodeconfigid, it will query the database (but not update the cache)
+ */
+trait NodeConfigurationService {
+  /**
+   * retrieve expected reports by config version
+   */
+  def findNodeExpectedReports(
+         nodeConfigIds: Set[NodeAndConfigId]
+      ): Box[Map[NodeAndConfigId, Option[NodeExpectedReports]]]
```

Comment:
Can you return `IOResult` in place of Box` ? We try to get ride of them, and `IOResult` compose better